### PR TITLE
[fix][broker] Close transactionBuffer after MessageDeduplication#checkStatus failed

### PR DESCRIPTION
**Pull request by tjiuming**
### Motivation

Currently, we didn't close the `TransactionBuffer` after `MessageDeduplication#checkStatus` failed. see: https://github.com/apache/pulsar/blob/master/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java#L1600
it will lead to memory leak.
In `TopicTransactionBuffer`, it will create a `snapshotWrite` in [`SingleSnapshotAbortedTxnProcessorImpl`](https://github.com/apache/pulsar/blob/master/pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/SingleSnapshotAbortedTxnProcessorImpl.java#L57), and the write will held by [`SystemTopicClientBase#writers`](https://github.com/apache/pulsar/blob/master/pulsar-broker/src/main/java/org/apache/pulsar/broker/systopic/SystemTopicClientBase.java#L76).

If we don't close the `TransactionBuffer` after `MessageDeduplication#checkStatus` failed, the number of `TransactionBufferSnapshotWriter` held by `SystemTopicClientBase#writers` will keep increasing, and may lead to OOM in the finally.

In https://github.com/apache/pulsar/pull/15015 , it reuses the `TransactionBufferSnapshotWriter`, which means that the topics of a same namespace will use one single `TransactionBufferSnapshotWriter` instance, it greatly reduces the amount of memory leaks, but if we don't close the `TransactionBuffer`, there are still memory leak risks.

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository:  https://github.com/tjiuming/pulsar/pull/17

### CODE REVIEW

**Review by Technoboy- — APPROVED** [review]
LGTM

**Review by congbobo184 — COMMENTED** [review]
better to add a test for it

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java [hunk 1]
```diff
@@ -1599,6 +1599,12 @@ public void openLedgerComplete(ManagedLedger ledger, Object ctx) {
                                         .exceptionally((ex) -> {
                                             log.warn("Replication or dedup check failed."
                                                     + " Removing topic from topics list {}, {}", topic, ex);
+                                            persistentTopic.getTransactionBuffer()
+                                                    .closeAsync()
+                                                    .exceptionally(t -> {
+                                                        log.error("[{}] Close transactionBuffer failed", topic, t);
+                                                        return null;
```
Comment (yaalsn) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java:1606]: Do you know if we need to return here?

Comment (tjiuming) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/BrokerService.java:1606]: CompletableFuture#exceptionally() requires a `Function` arg
image
and `Function` requires a return value.
image

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/buffer/TopicTransactionBufferTest.java [hunk 2]
```diff
@@ -86,4 +98,50 @@ public void testTransactionBufferAppendMarkerWriteFailState() throws Exception {
         FieldUtils.writeField(persistentTopic.getManagedLedger(), "state", ManagedLedgerImpl.State.WriteFailed, true);
         txn.commit().get();
     }
+
+    @Test
+    public void testCheckDeduplicationFailedWhenCreatePersistentTopic() throws Exception {
```
Comment (codelipenghui) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/buffer/TopicTransactionBufferTest.java:103]: @tjiuming The test can also get passed with the code fix. Could you please check again?

Comment (tjiuming) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/transaction/buffer/TopicTransactionBufferTest.java:104]: @codelipenghui I've fixed the test, PTAL

**Review by codelipenghui — APPROVED** [review]

**Comment by tjiuming** [comment]
https://github.com/apache/pulsar/pull/19287
https://github.com/apache/pulsar/pull/19288
https://github.com/apache/pulsar/pull/19289

**Pull request merged**